In [32]:
import numpy as np
import pandas as pd
from clean import manifest_clinical_merge
from collections import Counter

In [2]:
pd.set_option('display.max_columns', 100)

# TLDR: Predicting Low Risk/ Not Low Risk using Gene Expression

Low Risk = 1; Standard - High Risk = 0

## Important Takeaway:

### Lasso Logistic Regression fit on all features (~180 important features)

### RF Classification fit on all features (~1100 important features)

### XGBoost Classification fit on all features (~300 important features)

### 11 ensemble gene ids considered important by all three models**

'ENSG00000099998.16', 'ENSG00000124491.14', 'ENSG00000125430.7', 'ENSG00000130649.8', 'ENSG00000151468.10', 'ENSG00000159708.16', 'ENSG00000163106.9', 'ENSG00000198842.8', 'ENSG00000228172.4', 'ENSG00000260182.1', 'ENSG00000266733.4'

# Data Prep

In [3]:
manifest_df = pd.read_csv('../Manifest/GCD_TARGET_Data_Manifest_AML_NBL_WT_RT.csv')

wt_disc_df = pd.read_excel('../Clinical_Data/TARGET_WT_ClinicalData_Discovery_20160714_public.xlsx')
aml_disc_df = pd.read_excel('../Clinical_Data/TARGET_AML_ClinicalData_20160714.xlsx')
nbl_disc_df = pd.read_excel('../Clinical_Data/TARGET_NBL_ClinicalData_20151124.xlsx')

WT_df = manifest_clinical_merge(manifest_df, wt_disc_df, 'TARGET-WT')
AML_df = manifest_clinical_merge(manifest_df, aml_disc_df, 'TARGET-AML')
NBL_df = manifest_clinical_merge(manifest_df, nbl_disc_df, 'TARGET-NBL')

/Users/david/statsbomb/RNAseq_Cancer_Biomarkers/scripts/clean.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  target_df['TARGET USI'] = target_df.loc[:, 'entity_submitter_id'].apply(lambda x: x[:16])


In [4]:
from clean import assay_transpose

In [353]:
assay_df = pd.read_csv('../Expn_Data/TARGET_NBL_AML_RT_WT_TMMCPM_log2_Norm_Counts.csv.zip')
assay_t_df = assay_transpose(assay_df)

In [6]:
from clean import assay_clinical_merge

In [7]:
AML_genes = assay_clinical_merge(assay_t_df, AML_df)
WT_genes = assay_clinical_merge(assay_t_df, WT_df)
NBL_genes = assay_clinical_merge(assay_t_df, NBL_df)

# Modeling on AML_genes

## Classification - predict low_risk / not_low_risk in AML patients

1) Variance comparison: Subset of patients under 03 09A classifcation - choosing 1000 genes of highest/lowest variance 
between labeled low_risk, not_low_risk group

2) Model performance when randomly sampling genes

3) Lasso logistic regression

4) Tree modeling with features considered important by Lasso

5) Modeling on all features with RF and XGBoost

In [46]:
from model_comp import data_prep_columns, model_prep, model_comp, model_prep_loc, xgboost_tuner, create_model_df
from sklearn.metrics import log_loss
from sklearn.model_selection import train_test_split
from sklearn.ensemble import  RandomForestClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier

# 1. Variance Comparison

### High Variance

In [215]:
df, a = data_prep_columns(AML_genes, 'Max')
X_train, X_test, y_train, y_test, holdout, X = model_prep(df, a)
xg1, rf1 = model_comp(X_train, X_test, y_train, y_test)

/Users/david/statsbomb/RNAseq_Cancer_Biomarkers/scripts/model_comp.py:100: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  data['label'] = y.copy()
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.28771180389607903
Gradient Boost Log Loss 0.36210740308746076
Random Forest Log Loss 0.3262005795045274
XGBoost Log Loss 0.22184269636164647


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


In [216]:
df.head()

TARGET USI Diagnostic ID  Gender                       Race  \
3   TARGET-20-PADZCG           09A  Female                    Unknown   
5   TARGET-20-PAEAKL           09A  Female  Black or African American   
9   TARGET-20-PAEERJ           09A  Female                      White   
10  TARGET-20-PAEFGR           09A  Female                      White   
12  TARGET-20-PAEFGT           03A  Female                      White   

                 Ethnicity  Age at Diagnosis in Days First Event  \
3   Not Hispanic or Latino                      5325     Relapse   
5   Not Hispanic or Latino                      5218     Relapse   
9   Not Hispanic or Latino                      5082     Relapse   
10  Not Hispanic or Latino                       549     Relapse   
12  Not Hispanic or Latino                      5452     Relapse   

    Event Free Survival Time in Days Vital Status  \
3                                314         Dead   
5                                134         Dead   
9                                259         Dead   
10                               497        Alive   
12                               213         Dead   

    Overall Survival Time in Days  Year of Diagnosis  Year of Last Follow Up  \
3                             536               1996                    1998   
5                             163               1997                    1997   
9                             373               1997                    1998   
10                           4022               1997                    2008   
12                            315               1997                    1998   

    Protocol  WBC at Diagnosis Bone marrow leukemic blast percentage (%)  \
3   CCG-2961              29.3                                        58   
5   CCG-2961              96.1                                        80   
9   CCG-2961              71.9                                        60   
10  CCG-2961              51.2                                        75   
12  CCG-2961              11.4                                       100   

   Peripheral blasts (%) CNS disease Chloroma FAB Category   t(6;9)  t(8;21)  \
3                      5          No       No           M4       No       No   
5                     81         Yes       No           M5       No       No   
9                     41          No       No           M4       No       No   
10                    41          No       No           M2       No       No   
12                    70          No       No           M2  Unknown  Unknown   

   t(3;5)(q25;q34) t(6;11)(q27;q23) t(9;11)(p22;q23) t(10;11)(p11.2;q23)  \
3               No               No               No                  No   
5               No               No              Yes                  No   
9              Yes               No               No                  No   
10              No               No               No                  No   
12         Unknown          Unknown          Unknown             Unknown   

   t(11:19)(q23:p13.1)  inv(16)    del5q    del7q    del9q monosomy 5  \
3                   No       No       No       No       No         No   
5                  Yes       No       No       No       No         No   
9                   No       No       No       No       No         No   
10                  No      Yes       No       No       No         No   
12             Unknown  Unknown  Unknown  Unknown  Unknown    Unknown   

   monosomy 7 trisomy 8 trisomy 21      MLL  Minus Y  Minus X  \
3          No        No         No       No       No       No   
5          No        No         No      Yes       No       No   
9          No        No         No       No       No       No   
10         No        No         No       No       No       No   
12    Unknown   Unknown    Unknown  Unknown  Unknown  Unknown   

   Cytogenetic Code Other Cytogenetic Complexity Primary Cytogenetic Code  \
3                                              0            

In [218]:
xg1_important = X.columns[np.array(xg1.feature_importances_) > 0]
rf1_important = X.columns[np.array(rf1.feature_importances_) > 0]

In [219]:
print(len(xg1_important))
len(rf1_important)

205


697

### country frequency of genes being labeled 'important' by RF or XGboost

In [35]:
# xg_gene_counter = Counter()
# rf_gene_counter = Counter()

In [36]:
# for gene in xg1_important:
#     xg_gene_counter[gene] +=1
# for gene in rf1_important:
#     rf_gene_counter[gene] +=1

### 1b. low variance genes between two labeled groups

In [220]:
df, a = data_prep_columns(AML_genes, 'Min')
X_train, X_test, y_train, y_test, holdout, X = model_prep(df, a)
xg2, rf2 = model_comp(X_train, X_test, y_train, y_test)

/Users/david/statsbomb/RNAseq_Cancer_Biomarkers/scripts/model_comp.py:100: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  data['label'] = y.copy()
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.4236865689681798
Gradient Boost Log Loss 0.49256451159202913
Random Forest Log Loss 0.48223471620602953
XGBoost Log Loss 0.3429010650394736


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


In [221]:
xg2_important = X.columns[np.array(xg2.feature_importances_) > 0]
rf2_important = X.columns[np.array(rf2.feature_importances_) > 0]

In [223]:
print(len(xg2_important))
len(rf2_important)

208


766

In [222]:
# for gene in xg2_important:
#     xg_gene_counter[gene] +=1
# for gene in rf2_important:
#     rf_gene_counter[gene] +=1

# 2.random gene selection

- random_cols_ are **column index numbers** chosen at random

- does not perform as well as the variation selected columns

- use **model_prep** instead of **model_prep_loc** as random_columns_ is a randomly generated list of numbers 
used to select columns

In [230]:
random_cols = np.random.choice(21404, 2000)
random_set = set(random_cols)
random_cols = list(random_set)
random_cols_ = [x + 84 for x in random_cols]

In [231]:
# random_cols_ are column index numbers

In [232]:
len(random_cols_)

1914

In [233]:
df, __ = data_prep_columns(AML_genes, 'neither')
X_train, X_test, y_train, y_test, holdout, X = model_prep(df, random_cols_)
xg_random, rf_random = model_comp(X_train, X_test, y_train, y_test)

/Users/david/statsbomb/RNAseq_Cancer_Biomarkers/scripts/model_comp.py:100: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  data['label'] = y.copy()
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.39491046043536376
Gradient Boost Log Loss 0.4643323954434488
Random Forest Log Loss 0.4528452946267843
XGBoost Log Loss 0.3226523800073443


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


In [228]:
random_cols = np.random.choice(21404, 2000)
random_set = set(random_cols)
random_cols = list(random_set)
random_cols_ = [x + 84 for x in random_cols]
df, __ = data_prep_columns(AML_genes, 'neither')
X_train, X_test, y_train, y_test, holdout, X = model_prep(df, random_cols_)
xg_random2, rf_random2 = model_comp(X_train, X_test, y_train, y_test)

/Users/david/statsbomb/RNAseq_Cancer_Biomarkers/scripts/model_comp.py:100: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  data['label'] = y.copy()
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.3857378566561362
Gradient Boost Log Loss 0.450060344474042
Random Forest Log Loss 0.42749734747068563
XGBoost Log Loss 0.3699347494119728


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


In [234]:
random_cols = np.random.choice(21404, 2000)
random_set = set(random_cols)
random_cols = list(random_set)
random_cols_ = [x + 84 for x in random_cols]
df, __ = data_prep_columns(AML_genes, 'neither')
X_train, X_test, y_train, y_test, holdout, X = model_prep(df, random_cols_)
xg_random3, rf_rando3 = model_comp(X_train, X_test, y_train, y_test)

/Users/david/statsbomb/RNAseq_Cancer_Biomarkers/scripts/model_comp.py:100: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  data['label'] = y.copy()
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.3772384943194
Gradient Boost Log Loss 0.504297547416616
Random Forest Log Loss 0.4216150371839307
XGBoost Log Loss 0.3159291592955187


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.2s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


# 3. Lasso Logistic Regression using all columns

- lasso reduces to 3212 genes instead of 20,000+

**see bottom to see fully converged lasso**

In [69]:
from sklearn.linear_model import LogisticRegression

In [235]:
lasso_columns = df.iloc[:, 84:-4].columns
df, __ = data_prep_columns(AML_genes, 'neither')
X_train, X_test, y_train, y_test, holdout, X = model_prep_loc(df, lasso_columns)
log_model = LogisticRegression(penalty='l1', solver='saga')
log_model.fit(X_train, y_train)
y_pred = log_model.predict_proba(X_test)
log_loss(y_test, y_pred)

/Users/david/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/sag.py:334: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


0.21263654943160015

In [236]:
log_coefs = np.array(log_model.coef_)
mask = log_coefs > 0
important_lasso_genes = lasso_columns[mask[0]]

In [237]:
len(important_lasso_genes)

3110

In [ ]:
# for gene in important_lasso_genes:
#     xg_gene_counter[gene] +=1
#     rf_gene_counter[gene] +=1

### 4. Trees Performance Using Lasso Genes

In [238]:
df, __ = data_prep_columns(AML_genes, 'neither')
X_train, X_test, y_train, y_test, holdout, X = model_prep_loc(df, important_lasso_genes)
xg_lass, rf_lass = model_comp(X_train, X_test, y_train, y_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.38889814893691244
Gradient Boost Log Loss 0.4288557996806494
Random Forest Log Loss 0.40714678601797455
XGBoost Log Loss 0.38903383744528164


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.3s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


In [239]:
xg_lass_important = X.columns[np.array(xg_lass.feature_importances_) > 0]
rf_lass_important = X.columns[np.array(rf_lass.feature_importances_) > 0]

In [86]:
xgboost_tuner(X_train, X_test, y_train, y_test, [4, 6, 8, 10, 12])

XGB n_estimators: 400 log loss 0.4201527334044914
XGB n_estimators: 600 log loss 0.3856683401894328
XGB n_estimators: 800 log loss 0.3854842494317406
XGB n_estimators: 1000 log loss 0.3859489940459261
XGB n_estimators: 1200 log loss 0.3860728672479053


# 5. Feature Importance When Running Tree Over All Features

In [344]:
all_columns = df.iloc[:, 84:-4].columns
df, __ = data_prep_columns(AML_genes, 'neither')
X_train, X_test, y_train, y_test, holdout, X = model_prep_loc(df, all_columns)
xg_all, rf_all = model_comp(X_train, X_test, y_train, y_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Ensemble Log Loss 0.3417890185477423
Gradient Boost Log Loss 0.4407397864037863
Random Forest Log Loss 0.39492259532890717
XGBoost Log Loss 0.25292958437490304


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.6s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    0.0s finished


In [345]:
xg_all_important = X.columns[np.array(xg_all.feature_importances_) > 0]
rf_all_important = X.columns[np.array(rf_all.feature_importances_) > 0]

In [346]:
print(len(xg_all_important))
len(rf_all_important)

332


1147

In [348]:
# holdout.iloc[:, :-1].head()

In [349]:
# holdout.iloc[:, :-1].info()

### Performance with holdout set

In [350]:
p_holdout_pred =  xg_all.predict_proba(holdout.iloc[:, :-1].values)

In [351]:
log_loss(holdout.iloc[:, -1], p_holdout_pred)

0.18713662400841713

# Comparing the features that the models think are important

### Lasso vs. XGBoost vs. RF (on all columns)

In [243]:
print(len(important_lasso_genes))
print(len(xg_all_important))
len(rf_all_important)

3110
332


1147

In [244]:
set_lasso = set(important_lasso_genes)
set_xg = set(xg_all_important)
set_rf = set(rf_all_important)

In [245]:
print(len(set(important_lasso_genes)))
print(len(set(xg_all_important)))
len(set(rf_all_important))

3110
332


1147

Genes deemed important by both **xg and rf**

In [246]:
xg_rf_intersection = set_xg.intersection(set_rf)
len(xg_rf_intersection)

48

Genes deemed important by both **lasso and rf**

In [247]:
lasso_rf_intersection = set_lasso.intersection(set_rf)
len(lasso_rf_intersection)

290

Genes deemed important by both **lasso and xg**

In [248]:
lasso_xg_intersection = set_lasso.intersection(set_xg)
len(lasso_xg_intersection)

75

Intersection of ** lasso, xg and rf**

In [249]:
three_way = set.intersection(set_lasso, set_xg, set_rf)
len(three_way)

11

In [250]:
three_way

{'ENSG00000099998.16',
 'ENSG00000124491.14',
 'ENSG00000125430.7',
 'ENSG00000130649.8',
 'ENSG00000151468.10',
 'ENSG00000159708.16',
 'ENSG00000163106.9',
 'ENSG00000198842.8',
 'ENSG00000228172.4',
 'ENSG00000260182.1',
 'ENSG00000266733.4'}

### THESE ARE OUR MOST IMPORTANT GENE ENSEMBLES

'ENSG00000099998.16',
 'ENSG00000124491.14',
 'ENSG00000125430.7',
 'ENSG00000130649.8',
 'ENSG00000151468.10',
 'ENSG00000159708.16',
 'ENSG00000163106.9',
 'ENSG00000198842.8',
 'ENSG00000228172.4',
 'ENSG00000260182.1',
 'ENSG00000266733.4'

# Making a table of genes 

In [106]:
all_genes = set.union(set_lasso, set_xg, set_rf)
len(all_genes)

4285

In [ ]:
# test1 = ['a', 'b', 'c']
# test2 = ['d', 'c', 'e']
# test_list ['a', 'b', 'c']
# test_dict = {}
# for letter in test_list:
#     if item in 



In [107]:
# gene_dict = {}
# for gene in all_genes:
#     if gene in three_way:
#         gene_dict[gene] = 'RF, XGBoost, Lasso'
#     elif gene in lasso_xg_intersection:
#         gene_dict[gene] = 'XGBoost, Lasso'
#     elif gene in lasso_rf_intersection:
#         gene_dict[gene] = 'RF, Lasso'
        

SETS IM WORKING WITH:


-  three_way (union of all 11)

- lasso_xg_intersection, xg_rf_intersection, lasso_rf_intersection

- set_lasso, set_xg, set_rf (individual)

In [115]:
gene_dict = {}
for gene in all_genes:
    if gene not in gene_dict:
        gene_dict[gene] = []
        if gene in set_lasso:
            gene_dict[gene].append('lasso')
        if gene in set_xg:
            gene_dict[gene].append('xgboost')
        if gene in set_rf:
            gene_dict[gene].append('random forest')
    else:
        if gene in set_lasso:
            gene_dict[gene].append('lasso')
        if gene in set_xg:
            gene_dict[gene].append('xgboost')
        if gene in set_rf:
            gene_dict[gene].append('random forest')

In [180]:
gene_dict = {}
for gene in all_genes:
    if gene not in gene_dict:
        gene_dict[gene] = ''
        if gene in set_lasso:
            gene_dict[gene] += 'lasso '
        if gene in set_xg:
            gene_dict[gene] += 'xgb '
        if gene in set_rf:
            gene_dict[gene] += 'rf '
    else:
        if gene in set_lasso:
            gene_dict[gene] += ' lasso '
        if gene in set_xg:
            gene_dict[gene] += ' xgb '
        if gene in set_rf:
            gene_dict[gene] += ' rf '

In [181]:
len(gene_dict)

4285

In [182]:
gene_dict['ENSG00000259321.1']

'lasso rf '

In [183]:
d = list(gene_dict.values())

In [184]:
important_genes = pd.DataFrame(data = d, index=gene_dict.keys(), columns = ['Model'])

In [288]:
important_genes.head(5)

Model
ENSG00000173597.7   lasso 
ENSG00000143815.13     rf 
ENSG00000110583.11     rf 
ENSG00000235776.2      rf 
ENSG00000104213.11  lasso

In [ ]:
# if gene in set_lasso:
#             gene_dict[gene] += ' lasso '
#         if gene in set_xg:
#             gene_dict[gene] += ' xgb '
#         if gene in set_rf:
#             gene_dict[gene] += ' rf '

In [186]:
lasso_dict = {}
for gene in all_genes:
    if gene in set_lasso:
        lasso_dict[gene] = 1
    else:
        lasso_dict[gene] = 0

rf_dict = {}
for gene in all_genes:
    if gene in set_rf:
        rf_dict[gene] = 1
    else:
        rf_dict[gene] = 0
        
xgb_dict = {}
for gene in all_genes:
    if gene in set_xg:
        xgb_dict[gene] = 1
    else:
        xgb_dict[gene] = 0

In [190]:
a = list(lasso_dict.values())
b = list(rf_dict.values())
c = list(xgb_dict.values())

In [210]:
pd.DataFrame.from_dict([lasso_dict, rf_dict, xgb_dict])

ENSG00000000003.13  ENSG00000002586.16  ENSG00000002726.18  \
0                   0                   1                   1   
1                   1                   0                   0   
2                   0                   0                   0   

   ENSG00000002933.6  ENSG00000003056.6  ENSG00000003137.7  \
0                  1                  0                  1   
1                  0                  0                  0   
2                  0                  1                  0   

   ENSG00000003249.12  ENSG00000003436.13  ENSG00000004660.13  \
0                   1                   1                   0   
1                   0                   0                   1   
2                   0                   0                   0   

   ENSG00000004939.12  ENSG00000005073.5  ENSG00000005102.11  \
0                   0                  0                   1   
1                   0                  1                   0   
2                   1                  0                   0   

   ENSG00000005108.14  ENSG00000005194.13  ENSG00000005243.8  \
0                   1                   0                  1   
1                   0                   1                  0   
2                   0                   0                  0   

   ENSG00000005381.7  ENSG00000005471.14  ENSG00000005513.9  \
0                  1                   0                  1   
1                  0                   1                  0   
2                  0                   0                  0   

   ENSG00000005884.16  ENSG00000005893.14  ENSG00000006025.10  \
0                   1                   0                   0   
1                   0                   1                   1   
2                   0                   0                   0   

   ENSG00000006042.10  ENSG00000006047.11  ENSG00000006118.13  \
0                   0                   1                   1   
1                   0                   0                   0   
2                   1                   0                   0   

   ENSG00000006283.16  ENSG00000006459.9  ENSG00000006534.14  \
0                   1                  1                   1   
1                   0                  0                   0   
2                   0                  0                   0   

   ENSG00000006611.14  ENSG00000006638.10  ENSG00000006747.13  \
0                   1                   1                   0   
1                   0                   1                   0   
2                   0                   0                   1   

   ENSG00000007001.11  ENSG00000007038.9  ENSG00000007129.16  \
0                   1                  1                   1   
1                   0                  0                   0   
2                   0                  0                   0   

   ENSG00000007237.17  ENSG00000007314.10  ENSG00000007350.15  \
0                   1                   1                   0   
1                   1                   0                   1   
2                   0                   0                   0   

   ENSG00000007402.10  ENSG00000007516.12  ENSG00000007866.17  \
0                   1                   1                   1   
1                   0                   1                   0   
2                   0                   0                   0   

   ENSG00000007944.13  ENSG00000008056.11  ENSG00000008283.14  \
0                   0                   1                   1   
1                   1                   0                   0   
2                   0                   0                   0   

   ENSG00000008311.13  ENSG00000008394.11  ENSG00000008516.15  \
0                   1                   1                   1   
1                   1                   0                   0   
2                   0                   0                   0   

   ENSG00000008517.15  ENSG00000009790.13  ENSG00000010319.5  \
0                   1                   1                  0

# DIRECTIONALITY

looking at 11 genes deemed important by all three models

In [262]:
def find_indices(lst, lst2):
    return [i for i, elem in enumerate(lst) if elem in lst2]

In [276]:
X.head()

ENSG00000000003.13 ENSG00000000005.5 ENSG00000000419.11 ENSG00000000457.12  \
5             -2.0565          -4.74701            4.01127             1.7086   
9           -0.600677          -6.92715            4.33729            3.70407   
14           -1.44812          -5.93369            4.49801            1.87355   
17          -0.850633          -6.92715             5.0871             2.0163   
19           -1.49953          -6.92715            4.56887             2.5829   

   ENSG00000000460.15 ENSG00000000938.11 ENSG00000000971.14  \
5             2.45773            8.33617          -0.390612   
9             1.45783            5.80541            3.08307   
14            1.67079            6.36209           -2.93944   
17            2.02289            5.48147            1.12714   
19            2.74927            8.25123           -3.69325   

   ENSG00000001036.12 ENSG00000001084.9 ENSG00000001167.13 ENSG00000001460.16  \
5             5.86531           2.81586             4.3824             2.7385   
9             4.70263           5.93286            3.68418            1.49483   
14            5.49176           4.04131            5.00884          -0.132792   
17            3.71878           5.79254            5.24197          0.0510899   
19             6.3299           3.71387            5.55935            1.96017   

   ENSG00000001461.15 ENSG00000001497.15 ENSG00000001561.6 ENSG00000001617.10  \
5             2.93014            5.38245          0.610766        -0.00724986   
9             3.81156            4.45815           5.90521            2.60909   
14            2.92239            4.47019           5.27709            1.57547   
17            2.13054            4.82633           4.86902            2.25889   
19            4.07961            4.95494           3.40502           0.460694   

   ENSG00000001626.13 ENSG00000001629.8 ENSG00000001630.14 ENSG00000001631.13  \
5            -6.92715            2.7025           0.592369            2.33861   
9            -6.92715           3.98982           0.267381            3.63427   
14           -6.92715           4.28916            1.76021            3.41949   
17           -6.92715           3.51067            1.36873            3.61357   
19           -6.92715           3.92808            2.43959            2.50427   

   ENSG00000002016.15 ENSG00000002330.12 ENSG00000002549.11  \
5             2.51256            4.33478            5.21782   
9             1.96923            3.15996            5.24463   
14            2.13122            3.20822            5.70585   
17             2.6879            3.04914            6.23952   
19            1.30324            4.47159            6.52478   

   ENSG00000002586.16 ENSG00000002587.8 ENSG00000002726.18 ENSG00000002746.13  \
5             8.32612          -2.53384            1.66084           -1.48807   
9             8.50432          -1.56366         -0.0547583           -5.86665   
14             8.3289          -1.54763           0.659716           -5.93369   
17            6.58627           2.63941           -3.97085           -4.46574   
19            8.39484           -1.7447          -0.232569           -6.92715   

   ENSG00000002822.14 ENSG00000002834.16 ENSG00000002919.13 ENSG00000002933.6  \
5             4.79562            6.58635            4.58207           3.27681   
9              2.9938            7.11946            4.06171         -0.879721   
14            3.81296            6.66139             4.0911           4.97972   
17            3.66081            7.16637            4.00996           3.05879   
19            4.74647            6.82597            4.52192           4.64733   

   ENSG00000003056.6 ENSG00000003096.12 ENSG00000003137.7 ENSG00000003147.16  \
5            6.57761           0.929309          -5.18099            1.85467   
9            5.81253           0.494469          -4.50961            3.90215   
14           6.08954            2.85579          -4.13103          0.0710648   
17 

In [284]:
direct_mask = find_indices(X.columns, three_way)

In [285]:
direct_mask

[2046, 5133, 5227, 5805, 8576, 9431, 9905, 15011, 16650, 19201, 19735]

In [286]:
log_model.coef_[0][direct_mask]

array([0.00293309, 0.00267322, 0.0011049 , 0.00313885, 0.00027107,
       0.0018612 , 0.00510481, 0.00475758, 0.00019398, 0.0070231 ,
       0.00383056])

# Make lasso converge - takeaways:

- second lasso took 3212 features --> 174

- intersection between new lasso (174) and XGboost (332) --> 15 in common

- of those 15 - 6 were in the original 11, 9 new

In [291]:
lasso_columns = df.iloc[:, 84:-4].columns
df, __ = data_prep_columns(AML_genes, 'neither')
X_train2, X_test2, y_train2, y_test2, holdout, X = model_prep_loc(df, lasso_columns)
log_model2 = LogisticRegression(penalty='l1', solver='saga', max_iter=10000)
log_model2.fit(X_train2, y_train2)
y_pred2 = log_model2.predict_proba(X_test2)
log_loss(y_test2, y_pred2)

0.18823188523395454

In [292]:
log_coefs2 = np.array(log_model2.coef_)
mask2 = log_coefs2 > 0
important_lasso_genes2 = lasso_columns[mask2[0]]
len(important_lasso_genes2)

174

In [314]:
holdout.head()

ENSG00000000003.13 ENSG00000000005.5 ENSG00000000419.11  \
139           -1.44173          -6.92715            5.38677   
12            0.104784          -6.92715            4.81266   
38            -5.61165          -6.92715            3.90408   
48            -5.25508          -6.92715            3.23629   
106            1.95826          -6.92715            4.44381   

    ENSG00000000457.12 ENSG00000000460.15 ENSG00000000938.11  \
139            2.49248            2.93061            5.05417   
12             2.55302           0.438541            3.83624   
38             2.64333            2.25689             8.4203   
48             1.91899            2.19289            2.72304   
106            2.74641            2.49835            1.97342   

    ENSG00000000971.14 ENSG00000001036.12 ENSG00000001084.9  \
139           -1.15394            4.67788           4.36731   
12             1.24481            3.59023           4.07956   
38             -2.2936            5.92398           3.23892   
48           -0.571218            4.72464           3.23904   
106           0.317393            4.18695           4.42188   

    ENSG00000001167.13 ENSG00000001460.16 ENSG00000001461.15  \
139            5.16476            1.21325             3.7806   
12             3.15866          -0.179674            2.51379   
38             2.79599            3.40475            3.68965   
48             3.55356            2.35874            3.70236   
106             5.0975           -0.60401            2.52431   

    ENSG00000001497.15 ENSG00000001561.6 ENSG00000001617.10  \
139            4.44626           4.45698           0.797512   
12             4.11495           5.30347            2.16562   
38             5.34347           2.83362            4.78065   
48             5.29401           1.05282            1.43482   
106            5.86077           2.37444           -0.81346   

    ENSG00000001626.13 ENSG00000001629.8 ENSG00000001630.14  \
139           -5.65638           4.30682              1.524   
12            -6.92715           2.90603         -0.0176145   
38            -6.92715           1.56994          -0.406389   
48            -6.92715           2.62139            1.27261   
106            -6.0535           3.98804            1.34794   

    ENSG00000001631.13 ENSG00000002016.15 ENSG00000002330.12  \
139            4.15432            2.93061            3.08466   
12             3.56699            2.66071            3.71078   
38             1.02794           0.929438            6.28928   
48             1.92583            3.02324            4.33277   
106            3.29604            2.24929            3.33274   

    ENSG00000002549.11 ENSG00000002586.16 ENSG00000002587.8  \
139            5.09395            6.13487          -1.53572   
12             4.40671            8.26942          -1.88564   
38             5.00335            9.33731          -3.07916   
48             4.30549            9.73869          -4.00874   
106            6.32064            8.43537           3.81867   

    ENSG00000002726.18 ENSG00000002746.13 ENSG00000002822.14  \
139            1.34437           -4.99139             3.4134   
12            -2.93597            -4.0422            2.30493   
38             5.74993           -4.12899             6.0315   
48             1.51863           -6.92715            4.08342   
106           -5.12109           -3.84129            3.46282   

    ENSG00000002834.16 ENSG00000002919.13 ENSG00000002933.6 ENSG00000003056.6  \
139            6.26471            3.51799          -1.74436           5.17268   
12             7.25928            3.83977          -2.93597           5.44518   
38             6.58457            4.96882           2.82617           6.64543   
48              6.0039            3.09174           2.36378           5.19062   
106            6.91687            3.88362         -0.884542           5.33186   

    ENSG00000003096.12 ENSG00000003137.7 ENSG00000003147.16  \
139            -1.8

In [316]:
y_holdout_pred =  log_model2.predict_proba(holdout.iloc[:, :-1])
log_loss(holdout.iloc[:, -1], y_holdout_pred)

0.17037272553094038

In [312]:
pd.DataFrame(data = important_lasso_genes2).to_csv('lasso_genes_178.csv')

In [310]:
important_lasso_genes2

Index(['ENSG00000005108.14', 'ENSG00000005513.9', 'ENSG00000019169.10',
       'ENSG00000067798.12', 'ENSG00000069535.13', 'ENSG00000072694.17',
       'ENSG00000079102.15', 'ENSG00000080493.12', 'ENSG00000087245.11',
       'ENSG00000095917.12',
       ...
       'ENSG00000272217.1', 'ENSG00000272342.1', 'ENSG00000272973.1',
       'ENSG00000273355.1', 'ENSG00000273449.1', 'ENSG00000273628.1',
       'ENSG00000274191.1', 'ENSG00000275769.1', 'ENSG00000279717.1',
       'ENSG00000280081.2'],
      dtype='object', length=174)

In [294]:
# set_xg

In [296]:
lasso_set2 = set(important_lasso_genes2)

In [300]:
xgb_lasso_set2 = lasso_set2.intersection(set_xg)
len(xgb_lasso_set2)

15

In [301]:
three_way

{'ENSG00000099998.16',
 'ENSG00000124491.14',
 'ENSG00000125430.7',
 'ENSG00000130649.8',
 'ENSG00000151468.10',
 'ENSG00000159708.16',
 'ENSG00000163106.9',
 'ENSG00000198842.8',
 'ENSG00000228172.4',
 'ENSG00000260182.1',
 'ENSG00000266733.4'}

In [302]:
xgb_lasso_set2

{'ENSG00000095917.12',
 'ENSG00000111341.8',
 'ENSG00000115112.7',
 'ENSG00000124491.14',
 'ENSG00000130649.8',
 'ENSG00000163106.9',
 'ENSG00000179639.9',
 'ENSG00000182700.4',
 'ENSG00000185345.17',
 'ENSG00000198842.8',
 'ENSG00000255882.1',
 'ENSG00000257207.5',
 'ENSG00000260182.1',
 'ENSG00000266733.4',
 'ENSG00000267453.5'}

In [303]:
xgb_lasso_set2.intersection(three_way)

{'ENSG00000124491.14',
 'ENSG00000130649.8',
 'ENSG00000163106.9',
 'ENSG00000198842.8',
 'ENSG00000260182.1',
 'ENSG00000266733.4'}

In [304]:
xgb_lasso_set2.difference(three_way)

{'ENSG00000095917.12',
 'ENSG00000111341.8',
 'ENSG00000115112.7',
 'ENSG00000179639.9',
 'ENSG00000182700.4',
 'ENSG00000185345.17',
 'ENSG00000255882.1',
 'ENSG00000257207.5',
 'ENSG00000267453.5'}